File: libparabolic/src/models/historicdownload.cpp
#include "models/historicdownload.h"

namespace Nickvision::TubeConverter::Shared::Models
{
    HistoricDownload::HistoricDownload(const std::string& url)
        : m_url{ url },
        m_dateTime{ boost::posix_time::second_clock::universal_time() }
    {

    }

    HistoricDownload::HistoricDownload(const std::string& url, const std::string& title, const std::filesystem::path& path)
        : m_url{ url },
        m_title{ title },
        m_path{ path },
        m_dateTime{ boost::posix_time::second_clock::universal_time() }
    {

    }

    HistoricDownload::HistoricDownload(const std::string& url, const std::string& title, const std::filesystem::path& path, const boost::posix_time::ptime& dateTime)
        : m_url{ url },
        m_title{ title },
        m_path{ path },
        m_dateTime{ dateTime }
    {

    }

    const std::string& HistoricDownload::getUrl() const
    {
        return m_url;
    }

    void HistoricDownload::setUrl(const std::string& url)
    {
        m_url = url;
    }

    const std::string& HistoricDownload::getTitle() const
    {
        return m_title;
    }

    void HistoricDownload::setTitle(const std::string& title)
    {
        m_title = title;
    }

    const std::filesystem::path& HistoricDownload::getPath() const
    {
        return m_path;
    }

    void HistoricDownload::setPath(const std::filesystem::path& path)
    {
        m_path = path;
    }

    const boost::posix_time::ptime& HistoricDownload::getDateTime() const
    {
        return m_dateTime;
    }

    void HistoricDownload::setDateTime(const boost::posix_time::ptime& dateTime)
    {
        m_dateTime = dateTime;
    }

    bool HistoricDownload::operator==(const HistoricDownload& other) const
    {
        return m_url == other.m_url;
    }

    bool HistoricDownload::operator!=(const HistoricDownload& other) const
    {
        return !operator==(other);
    }

    bool HistoricDownload::operator<(const HistoricDownload& other) const
    {
        return m_dateTime < other.m_dateTime;
    }

    bool HistoricDownload::operator>(const HistoricDownload& other) const
    {
        return m_dateTime > other.m_dateTime;
    }
}

File: Nickvision.Parabolic.Shared/Models/HistoricDownload.cs
using System;

namespace Nickvision.Parabolic.Shared.Models;

public class HistoricDownload
{
    public Uri Url { get; }
    public string Title { get; set; }
    public string Path { get; set; }
    public DateTime DownloadedOn { get; set; }

    public HistoricDownload(Uri url)
    {
        Url = url;
        Title = string.Empty;
        Path = string.Empty;
        DownloadedOn = DateTime.Now;
    }

    public int CompareTo(HistoricDownload? other)
    {
        if (other is null)
        {
            return 1;
        }
        return Url.ToString().CompareTo(other.Url.ToString());
    }

    public override bool Equals(object? obj) => obj is HistoricDownload other && Equals(other);

    public bool Equals(HistoricDownload? other) => other is not null && Url == other.Url;

    public override int GetHashCode() => Url.GetHashCode();

    public static bool operator >(HistoricDownload left, HistoricDownload right) => left.CompareTo(right) > 0;

    public static bool operator <(HistoricDownload left, HistoricDownload right) => left.CompareTo(right) < 0;

    public static bool operator >=(HistoricDownload left, HistoricDownload right) => left.CompareTo(right) >= 0;

    public static bool operator <=(HistoricDownload left, HistoricDownload right) => left.CompareTo(right) <= 0;

    public static bool operator ==(HistoricDownload left, HistoricDownload right) => left.Equals(right);

    public static bool operator !=(HistoricDownload left, HistoricDownload right) => !left.Equals(right);
}
